EIPS/eip-draft-whisper.md

```diff
@@ -0,0 +1,179 @@
+
+This packet contains two objects: integer (0x01) followed by a list of Whisper Envelopes.
+
+This packet is used for sending the standard Whisper envelopes.
+
+**PoW Requirement** [`2`, `PoW`]
+
+This packet contains two objects: integer (0x02) followed by a single floating point value of PoW.
+
+This packet is used by Whisper nodes for dynamic adjustment of their individual PoW requirements. Receipient of this message should no longer deliver the sender messages with PoW lower than specified in this message.
+
+PoW is defined as average number of iterations, required to find the current BestBit (the number of leading zero bits in the hash), divided by message size and TTL:
+
+	PoW = (2**BestBit) / (size * TTL)
+
+PoW calculation:
+
+	fn short_rlp(envelope) = rlp of envelope, excluding env_nonce field.
+	fn pow_hash(envelope, env_nonce) = sha3(short_rlp(envelope) ++ env_nonce)
+	fn pow(pow_hash, size, ttl) = 2**leading_zeros(pow_hash) / (size * ttl)
+
+where size is the size of the full RLP-encoded envelope.
+
+**Bloom Filter** [`3`, `bytes`]
+
+This packet contains two objects: integer (0x03) followed by a byte array of arbitrary size.
+
+This packet is used by Whisper nodes for sharing their interest in messages with specific topics.
+
+The Bloom filter is used to identify a number of topics to a peer without compromising (too much) privacy over precisely what topics are of interest. Precise control over the information content (and thus efficiency of the filter) may be maintained through the addition of bits.
+
+Blooms are formed by the bitwise OR operation on a number of bloomed topics. The bloom function takes the topic and projects them onto a 512-bit slice. At most, three bits are marked for each bloomed topic.
+
+The projection function is defined as a mapping from a a 4-byte slice S to a 512-bit slice D; for ease of explanation, S will dereference to bytes, whereas D will dereference to bits.
+
+	LET D[*] = 0
+	FOREACH i IN { 0, 1, 2 } DO
+	LET n = S[i]
+	IF S[3] & (2 ** i) THEN n += 512
+	D[n] = 1
+	END FOR
+
+
+OPTIONAL
+
+**P2P Request** [`126`, `whisper_envelope`]
+
+This packet contains two objects: integer (0x7E) followed by a single Whisper Envelope.
+
+This packet is used for sending Dapp-level peer-to-peer requests, e.g. Whisper Mail Client requesting old messages from the Whisper Mail Server.
+
+**P2P Message** [`127`, `whisper_envelope`]
+
+This packet contains two objects: integer (0x7F) followed by a single Whisper Envelope.
+
+This packet is used for sending the peer-to-peer messages, which are not supposed to be forwarded any further. E.g. it might be used by the Whisper Mail Server for delivery of old (expired) messages, which is otherwise not allowed.
+
+
+### Whisper Envelope
+
+Envelopes are RLP-encoded structures of the following format:
+
+	[ Version, Expiry, TTL, Topic, Data, Nonce ]
+	
+`Version`: 1 byte. If Version is higher than current, envelope could not be decoded, and therefore only forwarded to the peers.
```

Comment:
hm, but since payload is general then version should be contained within there, rather than within the envelope. I was under the impression that `version` here was to signify the envelope format version (if extra fields were added or something), and then we can't always know if there will be a TTL/Expiry.